src/mlmc/quantity_concept.py

```diff
@@ -121,120 +373,594 @@ def mult_op(x, y):
-        self.q_list = args
-        self.shape = [len(self.q_list)] + args[0].shape
+    def size(self) -> int:
+        return len(self._times) * self._qtype.size()
 
+    def __getitem__(self, key):
+        if key not in self._times:
+            raise KeyError("Item " + str(key) + " was not found in TimeSeries" +
+                           ". Available items: " + str(list(self._times)))
 
+        q_type = copy.deepcopy(self._qtype)
+        position = self._times.index(key)
+        q_type.start = position * q_type.size()
+        return q_type
 
-class Field(QBase):
-    def __init__(self, *args: Tuple[str, QBase]):
-        self.quantity_dict = OrderedDict(*args)
 
+class FieldType(QType):
+    def __init__(self, args: List[Tuple[str, QType]], start=0):
+        """
+        QType must have same structure
+        :param args:
+        """
+        self._dict = dict(args)
+        self._qtype = args[0][1]
+        self.start = start
+        assert all(q_type == self._qtype for _, q_type in args)
+
+    def size(self) -> int:
+        return len(self._dict.keys()) * self._qtype.size()
+
+    def __getitem__(self, key):
+        if key not in self._dict:
+            raise KeyError("Key " + str(key) + " was not found in FieldType" +
+                           ". Available keys: " + str(list(self._dict.keys())))
+
+        q_type = copy.deepcopy(self._qtype)
+        position = list(self._dict.keys()).index(key)
+        q_type.start = position * q_type.size()
+        return q_type
+
+    def __copy__(self):
+        new = type(self)([(k, v) for k, v in self._dict.items()])
+        new.__dict__.update(self.__dict__)
+        return new
+
+
+class DictType(QType):
+    def __init__(self, args: List[Tuple[str, QType]]):
+        self._dict = dict(args)  # Be aware we it is ordered dictionary
+        self.start = 0
+
+        self._check_base_type()
+
+    def _check_base_type(self):
+        qtypes = list(self._dict.values())
+        for qtype in qtypes[1:]:
+            if not isinstance(qtype.base_qtype(), type(qtypes[0].base_qtype())):
+                raise TypeError("qtype {} has base QType {}, expecting {}. "
+                                "All QTypes must have same base QType, either SacalarType or BoolType".
+                                format(qtype, qtype.base_qtype(), qtypes[0].base_qtype()))
+
+    def base_qtype(self):
+        return list(self._dict.values())[0].base_qtype()
+
+    def size(self) -> int:
+        return int(np.sum(q_type.size() for _, q_type in self._dict.items()))
+
+    def get_qtypes(self):
+        return self._dict.values()
+
+    def replace_scalar(self, new_qtype):
+        for key, qtype in self._dict.items():
+            if isinstance(qtype, ScalarType):
+                self._dict[key] = new_qtype
+            else:
+                qtype.replace_scalar(new_qtype)
+
+    def __getitem__(self, key):
+        if key not in self._dict:
```

Comment:
Just catch exception or possibly leave the default exception. Do not list all keys in the exception, possibly just a few of them.